src/messaging/ExchangeContext.cpp

```diff
@@ -489,6 +490,9 @@ void ExchangeContext::NotifyResponseTimeout(bool aCloseIfNeeded)
     }
 #endif // CONFIG_DEVICE_LAYER && CHIP_CONFIG_ENABLE_ICD_SERVER
 
+    // Grab the value of WaitingForResponseOrAck() before we mess with our state.
+    bool mrpTimedOut = WaitingForResponseOrAck();
```

Comment:
Is we are still waiting for a response or ack here, doesn't that means that the MRP hasn't run out of retries yet and is not timed out?

Reply:
`WaitingForResponseOrAck()` is true here if we never got an ack.  MRP might have run out of retries (in which case `IsWaitingForAck()` will be false), or it might not have (if the response timeout was set below the MRP timeout).  Normally people won't do the latter, but it's a good point that they might do that, and then marking the session defunct here is arguably wrong.  In either case, calling the variable `mrpTimedOut` if `IsWaitingForAck()` is true is obviously wrong.

I guess the questions are:
1. What can we call `WaitingForResponseOrAck()` that makes it clearer what it really means?  It's a flag that gets cleared if we get _either_ a response (which must carry an ack) or an ack.. but the latter only if we get the ack while `IsWaitingForAck()` is true.
2. Should we mark the session defunct if the exchange timeout was set below the MRP timeout, so `IsWaitingForAck()` is still true here?

@jmartinez-silabs thoughts?

Reply:
The name `WaitingForResponseOrAck()` is good for me, it is clear. However the bool variable `mrpTimeout` is set to true when we are still waiting for an ack. The exchange context time out hit here but MRP clears the waiting for ack flag when it runs out of retries. 

Shouldn't it be `mrpTimeout = !WaitingForResponseOrAck()` ? 

But in your second point, I agree that if the context timeout hit, we should defunct it even if MRP still has some retries

Reply:
`mrpTimedOut` would actually correspond to `!IsWaitingForAck()`.  What we are really after, here, is "did we ever get an ack"; I will rename the boolean accordingly (and fix the test for it to be clearer).